source/opt/fold_spec_const_op_composite.cpp

```diff
@@ -0,0 +1,565 @@
+  }
+  return result;
+}
+
+std::tuple<std::unique_ptr<Value>, std::vector<uint32_t>>
+FoldSpecConstantOpAndCompositePass::GetBroadcastOperationResult(
+    ir::Instruction* inst) {
+  const uint32_t spec_opcode_count = 1;
+  uint32_t num_operands = inst->NumInOperands();
+  uint32_t operand_index = 1;
+  uint32_t operand_id = inst->GetSingleWordInOperand(operand_index);
+  auto iter = id_to_const_val_.find(operand_id);
+  if (iter == id_to_const_val_.end()) {
+    return std::make_tuple(nullptr, std::vector<uint32_t>());
+
+  } else if (iter->second->AsScalar()) {
+    // Get the value in raw words for all the operands, prepare for the
+    // operation.
+    std::vector<uint32_t> operand_values_in_raw_words;
+    operand_values_in_raw_words.reserve(num_operands - spec_opcode_count);
+    operand_values_in_raw_words.push_back(iter->second->AsScalar()->word);
+    for (uint32_t i = operand_index + 1; i < num_operands; i++) {
+      iter = id_to_const_val_.find(inst->GetSingleWordInOperand(i));
+      if (iter == id_to_const_val_.end())
+        return std::make_tuple(nullptr, std::vector<uint32_t>());
+      // All the operands should be scalar constants.
+      assert(iter->second->AsScalar() && "All operands must be in scalar type");
+      operand_values_in_raw_words.push_back(iter->second->AsScalar()->word);
+    }
+    uint32_t result_in_raw_word =
+        Operate(static_cast<SpvOp>(inst->GetSingleWordInOperand(0)),
+                std::move(operand_values_in_raw_words));
+    return std::make_tuple(
+        std::unique_ptr<Value>(new ScalarValue(result_in_raw_word)),
+        std::vector<uint32_t>());
+
+  } else {
+    assert(
+        AllInOperandValuesAreConstVectorsInSameLength(inst, id_to_const_val_) &&
+        "All the operands must be vector constants with same number of "
+        "components");
+    (void)static_cast<bool (*)(
+        ir::Instruction*,
+        const std::unordered_map<uint32_t, std::unique_ptr<Value>>&)>(
+        AllInOperandValuesAreConstVectorsInSameLength);
+
+    CompositeValue* first_operand_value = iter->second->AsComposite();
+    uint32_t num_dims =
+        static_cast<uint32_t>(first_operand_value->nested.size());
+    // Get the value in raw words for all the operands in each dimension. The
+    // result from each dimension should be aggregated to get the final result
+    // vector.
+    auto result = std::unique_ptr<Value>(new CompositeValue);
+    std::vector<uint32_t> new_component_ids;
+    for (uint32_t d = 0; d < num_dims; d++) {
+      std::vector<uint32_t> one_dimension_operand_values_in_raw_words;
+      one_dimension_operand_values_in_raw_words.reserve(num_operands -
+                                                        spec_opcode_count);
+      one_dimension_operand_values_in_raw_words.push_back(
+          id_to_const_val_[first_operand_value->nested[d]]->AsScalar()->word);
+      for (uint32_t i = operand_index + 1; i < num_operands; i++) {
```

Comment:
the first id operand has been used to get the dimensions (the number of components), so `i` starts from 2. Anyway, I should have the `operand_index++` when getting the `operand_id`, then the following loops can start with `operand_index` instead of `operand_index + 1`.